src/mlpack/methods/naive_bayes/naive_bayes_classifier_impl.hpp

```diff
@@ -315,18 +321,24 @@ void NaiveBayesClassifier<ModelMatType>::Classify(
 
   predictions.set_size(data.n_cols);
 
-  ModelMatType logLikelihoods;
+  ModelMatType logLikelihoods, predictionLogProb;
   LogLikelihood(data, logLikelihoods);
 
   // This will hold log(Prob(X)) for each point.
   arma::Col<ElemType> logProbX(data.n_cols);
+  predictionLogProb.set_size(arma::size(logLikelihoods));
+  double maxValue;
   for (size_t j = 0; j < data.n_cols; ++j)
   {
-    logProbX(j) = log(arma::accu(exp(logLikelihoods.col(j))));
-    logLikelihoods.col(j) -= logProbX(j);
+    // To prevent underflow in log of sum of exp of x operation (where x is a
+    // small negative value), we use logsumexp(x - max(x)) + max(x).
+    maxValue = arma::max(logLikelihoods.col(j));
+    logProbX(j) = log(arma::accu(exp(logLikelihoods.col(j) - maxValue))) +
+        maxValue;
+    predictionLogProb.col(j) = logLikelihoods.col(j) - logProbX(j);
```

Comment:
Looks like if we store the result into ``predictionProbs`` right away we don't need ``predictionLogProb``.

Reply:
Right, we could just do `predictionProbs.col(j) = arma::exp(logLikelihoods.col(j) - logProbX(j))`.  Also, there's no need to store `logProbX`, I think---we can just use a single value, or even inline the current computation of `logProbX(j)`.